Add support for account-based channel bans

progval opened the pull request:
using https://github.com/ircv3/ircv3-specifications/pull/464

progval committed:
Add IrcState.nicksToAccount

To keep track of network accounts using various IRCv3 specs

progval committed:
core & Channel: Add option --account to @kban and @iban

A future commit will add support for 'account' in
supybot.protocols.irc.banmask, but it is not supported for now,
as that config value is also used for ignore masks

progval committed:
Add support for 'account' in supybot.protocols.irc.banmask

And a new method .makeExtBanmask() as an alternative to .makeBanmask(),
so plugins can opt-in to extended banmasks when they support it.

'ignore' commands in Channel and anti-flood in Owner and Misc will
keep using .makeBanmask() because they use them as ignore masks in
ircdb.

progval committed:
Replace makeExtBanmask with makeExtBanmasks

Now that we can return both account extbans and regular masks,
it makes sense to ban both.

Otherwise, adding 'account' to supybot.protocols.irc.banmask
means we banned only the account instead of the hostmask,
which arguably makes the ban weaker (/NS LOGOUT to evade)

jlu5 commented on src/conf.py:
```diff
@@ -1215,33 +1215,87 @@ def makeBanmask(self, hostmask, options=None, channel=None, network=None):
+        print(masks)
+        return masks[0]
+
+    def makeExtBanmasks(self, hostmask, options=None, channel=None, *, network):
```
`s/options/banstyles/` perhaps?

progval replied:
it's meant to be existent with the existing `makeBanmask` :/

jlu5 replied:
Ah I missed that, not a big deal then

jlu5 commented on src/conf.py:
```diff
@@ -1215,33 +1215,87 @@ def makeBanmask(self, hostmask, options=None, channel=None, network=None):
+        masks = []
+
         for option in options:
             if option == 'nick':
```
I usually prefer Enum types instead of fixed string inputs, especially if we silently ignore invalid options.

progval replied:
ditto

progval replied:
I guess I can allow both and deprecate the string though.

progval replied:
Heh, I'll just log invalid options instead.

jlu5 commented on plugins/Channel/test.py:
```diff
@@ -161,9 +163,13 @@ def testVoice(self):
+        self.assertIn(m, [
+            ircmsgs.bans(self.channel, permutation)
+            for permutation in itertools.permutations(hostmasks)
+            ])
```
Would https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual work better here?

progval replied:
Heh, I don't know

```diff
diff --git a/plugins/Channel/test.py b/plugins/Channel/test.py
index bd716767c..8b08df6e3 100644
--- a/plugins/Channel/test.py
+++ b/plugins/Channel/test.py
@@ -165,10 +165,9 @@ def testVoice(self):
 
     def assertKban(self, query, *hostmasks, **kwargs):
         m = self.getMsg(query, **kwargs)
-        self.assertIn(m, [
-            ircmsgs.bans(self.channel, permutation)
-            for permutation in itertools.permutations(hostmasks)
-            ])
+        self.assertEqual(m.command, "MODE", m)
+        self.assertEqual(m.args[0], "+" + "b" * len(hostmasks), m)
+        self.assertCountEqual(m.args[1], hostmasks, m)
 
         m = self.getMsg(' ')
         self.assertEqual(m.command, 'KICK')
```

jlu5 commented on src/irclib.py:
```diff
@@ -680,6 +680,12 @@ class IrcState(IrcCommandDispatcher, log.Firewalled):
+
+    .. attribute:: nicksToAccounts
+
+        Stores the current network account name of a seen nick.
```
```suggestion
        Stores the current services account name of a seen nick.
```

jlu5 commented on plugins/Channel/test.py:
```diff
@@ -219,11 +225,88 @@ def join():
+                        'foobar!user@host.domain.tld')
+        join()
+        self.assertKban('kban --account foobar',
+                        'foobar!user@host.domain.tld')
```
I wonder if it'd be better to throw an error if only `--account` is given and the user's not logged in. Exact match bans are pretty trivial to evade on most networks (even accidentally if your connection times out), to the point I doubt there's much value in the bot setting them in the first place.

progval replied:
What if it uses the default banmask instead?

progval replied:
I made it fall back to the default banmask, then to just the host (in case the default banmask is only `['account']` for some reason)

jlu5 commented on src/irclib.py:
```diff
@@ -957,6 +979,11 @@ def do354(self, irc, msg):
         self.nicksToHostmasks[nick] = hostmask
+        if account == '0':
+            # logged out
+            self.nicksToAccounts[nick] = None
```
Storing None values seems a bit odd, would it be better to delete the entry?

progval replied:
no entry means "unknown", `None` entry means "logged out"

jlu5 replied:
It might be nice to have the function describe how it'll throw KeyError if the user isn't in the state

jlu5 commented on src/ircutils.py:
```diff
@@ -345,6 +345,25 @@ def banmask(hostmask):
             return '*!*@'  + host
 
+
+def accountExtban(nick, irc):
```
nit: having `irc` as the first argument is more consistent with the other functions here

progval replied:
Right. I probably wanted to make it optional or something, but I don't think it makes sense.

jlu5 commented on test/test_irclib.py:
```diff
@@ -579,6 +579,105 @@ def testChghost(self):
+        st.addMsg(self.irc, ircmsgs.join('#foo', prefix='bar!baz@qux'))
+        with self.assertRaises(KeyError):
+            st.nickToAccount('foo')
+            st.nickToAccount('bar')
```
Are we expecting both calls to raise, or one of the two?

jlu5 commented on test/test_irclib.py:
```diff
@@ -579,6 +579,105 @@ def testChghost(self):
+            with self.assertRaises(KeyError):
+                st2.nickToAccount('foo')
+            self.assertEqual(st2.nickToAccount('bar'), 'account2')
+            self.assertEqual(st.nickToAccount('foo'), 'account1')
```
Check `st2` instead of `st`?

progval replied:
nope, this is meant to check `st` isn't affected by a message added to `st2`

progval replied:
I added a comment to clear the confusion

jlu5 commented on test/test_irclib.py:
```diff
@@ -579,6 +579,105 @@ def testChghost(self):
+                st2.nickToAccount('foo')
+            self.assertEqual(st2.nickToAccount('foo2'), 'account1')
+            self.assertEqual(st2.nickToAccount('bar'), 'account2')
+            self.assertEqual(st.nickToAccount('foo'), 'account1')
```
Same as above

jlu5 commented on test/test_irclib.py:
```diff
@@ -579,6 +579,105 @@ def testChghost(self):
+            with self.assertRaises(KeyError):
+                st2.nickToAccount('foo')
+            self.assertEqual(st2.nickToAccount('bar'), 'account1')
+            self.assertEqual(st.nickToAccount('foo'), 'account1')
```
Same as above

jlu5 commented on src/conf.py:
```diff
@@ -1215,33 +1215,87 @@ def makeBanmask(self, hostmask, options=None, channel=None, network=None):
-            return hostmask
-        return ircutils.joinHostmask(bnick, buser, bhost)
+                ircutils.isUserHostmask(hostmask) and \
+                masks == []:
```
```suggestion
                not masks:
```

progval committed:
s/network/services/

Co-authored-by: James Lu <james@overdrivenetworks.com>

progval committed:
Explain why 'st' is checked too

progval committed:
Remove prints

progval committed:
Correctly document meaning of None vs absent values in nicksToAccounts

progval committed:
tests: Fix self.assertRaises(KeyError) for nickToAccount

progval committed:
s/masks == []/not masks/

progval committed:
Swap arguments of accountExtban

progval committed:
Avoid listing all permutations

progval committed:
makeExtBanmasks: Log invalid options

progval committed:
Cowardly refuse to ban oneself with an account extban

progval committed:
Merge branch 'master' into account-ban

progval committed:
Fix parenthesis in docstring

progval committed:
When only --account is provided, fallback to supybot.protocols.irc.banmask before exact mask

progval committed:
Fall back to banning host instead of exact mask

This only happens on the newly introduced account extban (in case the user
does not have an account, or the server does not provide accounts)
so this does not change existing behavior.

Falling back to the host instead of the exact mask makes it less easy
to evade these bans

progval committed:
Add test the bot won't account-extban itself

progval commented:
Should be all good now. The spec was just merged by IRCv3 (as a draft) so I'm going to merge this PR soon-ish
